Handle Emote Packets

HashimTheArab opened the pull request:
## Introduction
This pull request broadcasts emote packets to players and implements a cooldown with it so someone doesn't dos attack the server with emotes!

HashimTheArab committed:
handle emotes

dktapps committed:
bootstrap: check for zlib raw support in leveldb

HashimTheArab committed:
proper flags

Prim69 commented on src/network/mcpe/handler/InGamePacketHandler.php:
```diff
@@ -888,4 +889,12 @@ public function handleLevelSoundEvent(LevelSoundEventPacket $packet) : bool{
+
+	public function handleEmote(EmotePacket $packet) : bool {
+		if($this->player->getServer()->getTick() - $this->player->lastEmoteTick > 5){
+			$this->player->getWorld()->broadcastPacketToViewers($this->player->getPosition(), $packet);
```
Should be good now

dktapps commented on src/network/mcpe/handler/InGamePacketHandler.php:
```diff
@@ -888,4 +889,12 @@ public function handleLevelSoundEvent(LevelSoundEventPacket $packet) : bool{
+
+	public function handleEmote(EmotePacket $packet) : bool {
+		if($this->player->getServer()->getTick() - $this->player->lastEmoteTick > 5){
+			$this->player->getWorld()->broadcastPacketToViewers($this->player->getPosition(), EmotePacket::create($packet->getActorRuntimeId(), $packet->getEmoteId(), $packet::FLAG_SERVER));
```
```suggestion
			$this->player->getWorld()->broadcastPacketToViewers($this->player->getPosition(), EmotePacket::create($packet->getActorRuntimeId(), $packet->getEmoteId(), EmotePacket::FLAG_SERVER));
```
Don't use late static binding. It makes usage search harder and prevents optimisations.

dktapps commented:
We should also have an event for this.

HashimTheArab committed:
player emote event

Prim69 commented:
Anything else?

dktapps committed:
pocketmine/math 0.4.0

dktapps commented:
The build also needs to pass :)

HashimTheArab committed:
builds lmao

dktapps committed:
Release 4.0.0-BETA8

dktapps committed:
4.0.0-BETA9 is next

Prim69 commented:
Passed

dktapps committed:
World: remove spammy debug message

dktapps committed:
Arrow: do not add pickups to creative players' inventories
closes #2932

dktapps committed:
Cake: fixed desync on cancellation of eating
closes #3591
we don't support eating in creative right now, but the cake shouldn't appear to be eaten when it's not.

dktapps committed:
Remove useless newline

dktapps committed:
SurvivalBlockBreakHandler: remove useless code

dktapps committed:
World: Register a temporary chunk loader on chunks used by PopulationTask
fixes #3839

dktapps committed:
World: do not crash on duplicate tiles loaded from disk
closes #4049

dktapps committed:
World: avoid duplicated logger code in initChunk()

dktapps committed:
World::unregisterChunkListenerFromAll(): go through unregisterChunkListener()
this ensures that everything gets cleaned up properly (e.g. player chunk listeners).

dktapps committed:
changelog: mention block-picking changes
[ci skip]

dktapps committed:
World: eliminate final remaining 'no loaders attached' debug message on player creation

dktapps committed:
InGamePacketHandler: mark player as not using item in more cases
fixes #4355

dktapps committed:
Player: do not re-request the same ungenerated chunks multiple times
this doesn't affect chunk resends, since they'll be kicked back to NEEDED, which is detected by orderChunks().

dktapps committed:
Improved Promise API - separate resolver and consumer APIs
this makes creating a promise slightly more cumbersome, but I'm more concerned about people who might try to call 'new Promise' directly.

dktapps committed:
Advisory chunk locking for chunk population (#4513)

this allows chunks locked for population to be modified. If the PopulationTask detects that the chunk was modified during the onCompletion(), the result of the population will be discarded and rescheduled, so that it includes user modifications.

dktapps committed:
Revert "Player: do not re-request the same ungenerated chunks multiple times"

This reverts commit 866020dfdb06434b4f361a4a0f0faf7a89c30ed2.

For some fucking reason this broke resending chunks in some cases (and
sending chunks at all in others). I don't have time to debug this right
now, so it's going to have to remain broken, infuriatingly enough.

dktapps committed:
Promise: expose isResolved()

dktapps committed:
World::orderChunkPopulation() may return a pre-resolved promise
this does not indicate a failure; it indicates that the chunk has already been successfully populated.
In this case, we shouldn't be putting the task back on the queue.

This is skirting around the real bug, which is that requestChunkPopulation() doesn't check if the target chunk is already populated before it creates a new promise that it will be.

dktapps committed:
World: fixed original promise not getting fulfilled if the chunk became populated=true after a promise was already made (but not fulfilled) to populate it
this could happen if a plugin calls setPopulated(true) on a chunk after a request for its population landed in the queue, but before it actually got processed. In that case, the promise would never get fulfilled.

dktapps committed:
Revert "Revert "Player: do not re-request the same ungenerated chunks multiple times""

This reverts commit 3265d3f6b4d6705276b56fab5f765e2f97b5d992.

dktapps committed:
World: remove the fulfilled promise from the population request map
fixes crash when unregistering chunk loaders

dktapps committed:
World: fixed requestChunkPopulation() queuing requests for chunks which are already populated
this led to chunk sending getting bogged down if there were more than population-queue-size chunks waiting to be generated.

dktapps committed:
World: use better variable names in orderChunkPopulation()

dktapps committed:
World: check population locks _after_ checking if the chunk is already populated, not before
this led to another case where a population request would be queued up for an already-populated chunk for no reason.

dktapps committed:
World: flip orderChunkPopulation() condition around
this makes it more obvious that the code is similar to requestChunkPopulation() barring one distinct difference.

dktapps committed:
World: deduplicate code in request/orderChunkPopulation

dktapps committed:
World: avoid calling the same logic twice in requestChunkPopulation()
orderChunkPopulation() checks the preconditions too. Have them both call an internal function that doesn't.

dktapps committed:
Player: return from callback if used chunk status is not REQUESTED_GENERATION()
this can happen especially on large render distances when flying fast and changing direction - we decide we don't want the chunk, then, after changing direction and re-ordering chunks, we decide we do want it again, and end up registering a second callback. In this case, we need to ensure that only one of the callbacks gets executed (it doesn't matter which one).

dktapps committed:
World: fixed assertion failure when requesting, cancelling, and then re-requesting chunks via requestChunkPopulation()
if the request/cancel/re-request happens all in the time before the queue gets drained, chunk hashes may appear multiple times in the queue. We don't want to process them twice if this happens (although it's mostly harmless anyway).

dktapps committed:
World: remove premature optimisation of setBlockAt() introduced by ece28e5d7b0d4c8d12a9477cdd1d3337277e892d
closes #4531
it turns out that letting the light updates themselves handle this is faster than trying to get in the way.

dktapps commented:
I had a think about this and I don't like the logic being embedded in the packet handler like this. It should be possible for a plugin to make a player do an emote.

Player::emote(uuid) and NetworkSession::onEmote(player, uuid) should be added.

Prim69 commented:
Would NetworkSession::onEmote send a new emote packet from the given player to the owner of the network session? If that's the case that doesn't sound like the best solution because a new packet would be created for each player that it's broadcasted to

dktapps commented:
Yes. Stop premature optimising.

Prim69 commented:
What about this?

Player
```php
public function emote(string $emoteId) : void {
	$this->getServer()->broadcastPackets($this->getViewers(), [EmotePacket::create($this->getId(), $emoteId, EmotePacket::FLAG_SERVER)]);
	}
```

because otherwise it would be this

player
```php
public function doEmote(string $emoteId) : void {
	foreach($this->getServer()->getOnlinePlayers() as $player){
		$player->getNetworkSession()->onEmote($this, $emoteId);
	}
}
```
network session
```php
public function onEmote(Player $player, string $emoteId) : void {
	$this->sendDataPacket(EmotePacket::create($this->getId(), $emoteId, EmotePacket::FLAG_SERVER));
}
```

And I don't see NetworkSession::onEmote being used anywhere else so the first one looks like a better solution to me

dktapps commented:
Stop premature optimising.

dktapps commented:
Player is not supposed to interact with packets in any way. That's supposed to be handled exclusively by NetworkSession.

Right now the end goal hasn't been reached, but in the future these methods will be on an interface with multiple possible implementations (e.g. for multi version). Broadcasting packets gets in the way of that and it's going to go away in the future.

There are ways that the performance benefits can be retained for such cases, but for something as trivial as this it's premature optimisation - this is not going to be a hot path and nobody will notice either way. Nobody notices the overhead from creating `TextPacket` to broadcast for every player either, and that's a pathway that has an equivalent level of hotness.

Prim69 commented:
Ok that makes sense I didn't see any reason for the second solution before

HashimTheArab committed:
Merge branch 'master' of https://github.com/GrapeDevs/PocketMine-MP

HashimTheArab committed:
update emote handling

HashimTheArab committed:
Revert "update emote handling"

This reverts commit 50490b9f0749a4138f8ac4bc5463ee2b3d1a2340.

HashimTheArab committed:
revert phpstorm formatting the whole file and format specific parts

Prim69 commented:
Everythings set now except my git pull added a bunch of your commits to the pr, should I figure something out for that or will it be fine?

dktapps commented:
It's a GitHub bug. Those commits aren't actually part of the git diff.

The actual commit diff looks like this: 

![image](https://user-images.githubusercontent.com/14214667/139957128-40c67d9e-8df9-4fc3-a034-e5a098c425cc.png)

dktapps commented:
The event call and rate-limit check should also be done in `Player`. That way the `lastEmoteTick` doesn't need to be public also.

I also prefer `emote()` over `doEmote()` also.

HashimTheArab committed:
move things

HashimTheArab committed:
unused import

Prim69 commented:
Anything else?
